iNICIO CLASE COMPONENT

Dennettlander opened the pull request:
INICIO LA CLASE COMPONENT
- Añadida lista de Features a Component
- Añadida restricción de Features: mínimo 1, máximo 4 (pendiente de lanzar excepcion relacionada)
  -Añadida función para sumar una nueva Feature a un Component
  - Modificado el atributo Aspect de Feature para que pase de ser un objeto Aspect a un string

aleherse commented on spec/ComponentSpec.php:
```diff
@@ -0,0 +1,61 @@
+
+    function it_has_a_list_of_features(Feature $feature)
+    {
+        $this->featuresList()->shouldReturn(array($feature));
```
Aquí puedes usar la notación corta introducida en PHP 5.4 [$feature] en vez de array($feature)
Creo que usando el plural features, no necesitas añadir la palabra List

Dennettlander replied:
Hecho :)

Dennettlander replied:
Hecho!

aleherse commented on spec/ComponentSpec.php:
```diff
@@ -0,0 +1,61 @@
+
+    function it_has_between_1_and_4_features()
+    {
+        $this->featuresList()->shouldCountBetween(1, 4);
```
para que queda más claro el ejemplo, deberías comprobar que tal y como está construido devuelve uno.

``` php
$this->featuresList()->shouldHaveCount(1); 
```

Luego crearía un ejemplo con 4

``` php
function it_can_have_up_to_4_features(Feature $feature1, Feature $feature2, Feature $feature3, Feature $feature4)
{
  $this->beConstructedWith('Firebolt staff', [$feature1, $feature2, $feature3, $feature4]);

  $this->featuresList()->shouldHaveCount(4); 
}
```

Y ya el ejemplo de 5 que lance una excepción

Dennettlander replied:
Hecho!

aleherse commented on spec/ComponentSpec.php:
```diff
@@ -0,0 +1,61 @@
+{
+    function let(Feature $feature)
+    {
+        $this->beConstructedWith('Firebolt staff', $feature);
```
Creo que es mejor si se usa un array de features en el constructor en vez de un único valor

``` php
$this->beConstructedWith('Firebolt staff', [$feature]);
```

Y asegurarse de que lanza una excepción si se usan más de 4

Dennettlander replied:
Hecho!

aleherse commented on spec/ComponentSpec.php:
```diff
@@ -0,0 +1,61 @@
+        $this->featuresList()->shouldCountBetween(1, 4);
+    }
+
+    function it_can_adds_a_new_feature()
```
Yo utilizaría un mock en vez de crear una nueva feature, además que la cantidad sea uno se sabe que funciona de un ejemplo anterior

``` php
function it_can_adds_a_new_feature(Feature $newFeature)
{
  $this->add($newFeature);

  $this->featuresList()->shouldHaveCount(2);
}
```

Quizás en vez de llamar al método add puedes llamarlo addFeature para que quede más claro qué es lo que se añade

Commit:
Salvados los cambios anteriores a la crisis de floobits
INICIO LA CLASE COMPONENT

- Añadida lista de Features a Component
- Añadida restricción de Features: mínimo 1, máximo 4 (pendiente de lanzar excepcion relacionada)
-Añadida función para sumar una nueva Feature a un Component

 - Modificado el atributo Aspect de Feature para que pase de ser un objeto Aspect a un string

Commit:
añadidos floobits

Commit:
modifficado el constructor de component para que acepte arrays de features en lugar de features sueltas

Commit:
- añadidos métodos addFeature(), getFeature() y removeFeature()
- __constructor() ahora llama a addFeature()
- modificado array de clase de Component para que sea asociativo
- modificados tests para funcionar con arrays asociativos
- iniciado método destroyComponent()

Commit:
- modificado let() de ComponentSpec para que tenga features preconstruidas
- eliminado matcher
- iniciado metodo devastateComponent()

Commit:
-añadido metodo devastaste() a component
-añadida varible $devastated y metodo isDevastated a Component
-modificados nombres de test de Feature y Components referidos a flags

aleherse commented on spec/ComponentSpec.php:
```diff
@@ -0,0 +1,104 @@
+        $this->beConstructedWith('Firebolt Staff', [$feature1, $feature2]);
+        $this->getFeature('Rubí demoníaco')->shouldReturn($feature1);
+        $this->getFeature('Sortilegios protectores')->shouldReturn($feature2);
+        $this->getFeature('Hechizos malignos')->shouldReturn('That feature doesn\' exist');
```
en vez de devolver una cadena en este último caso debería lanzar una excepción y yo sacaría ese caso a otro método

aleherse commented on src/Component.php:
```diff
@@ -0,0 +1,120 @@
+            throw new InvalidArgumentException('Number of features will exceed 4');
+        }
+
+        $this->features[$newFeature->name()] = $newFeature;
```
añadir lógica en el caso de que se añada una feature a un componente devastado, no hacer nada o lanzar excepción dependiendo de lo estricto que quieras ser con los errores
